libminifi/include/core/flow/StructuredConfiguration.h

```diff
@@ -0,0 +1,227 @@
+
+  /**
+   * A helper function to parse the Properties Node for a processor.
+   *
+   * @param properties_node the Node containing the properties
+   * @param processor      the Processor to which to add the resulting properties
+   */
+  void parsePropertiesNode(const Node& properties_node, core::ConfigurableComponent& component, const std::string& component_name, const std::string& section);
+
+  /**
+   * Parses the Funnels section of a configuration.
+   * The resulting Funnels are added to the parent ProcessGroup.
+   *
+   * @param node   the Node containing the Funnels section
+   *                 of the configuration
+   * @param parent the root node of flow configuration to which
+   *                 to add the funnels that are parsed
+   */
+  void parseFunnels(const Node& node, core::ProcessGroup* parent);
+
+  /**
+   * A helper function for parsing or generating optional id fields.
+   *
+   * In parsing flow configurations for config schema v1, the
+   * 'id' field of most component types that contains a UUID is optional.
+   * This function will check for the existence of the specified
+   * idField in the specified node. If present, the field will be parsed
+   * as a UUID and the UUID string will be returned. If not present, a
+   * random UUID string will be generated and returned.
+   *
+   * @param node     a pointer to the Node that will be checked for the
+   *                   presence of an idField
+   * @param id_field  the string of the name of the idField to check for. This
+   *                   is optional and defaults to 'id'
+   * @return         the parsed or generated UUID string
+   */
+  std::string getOrGenerateId(const Node& node, const std::string& id_field = "id");
+
+  std::string getRequiredIdField(const Node& node, std::string_view section = "", std::string_view error_message = "");
+
+  /**
+   * This is a helper function for getting an optional value, if it exists.
+   * If it does not exist, returns the provided default value.
+   *
+   * @param node         the flow node to check
+   * @param field_name    the optional field key
+   * @param default_value the default value to use if field is not set
+   * @param section  [optional] the top level section of the config
+   *                       for the node. This is used fpr generating a
+   *                       useful info message for troubleshooting.
+   * @param info_message  [optional] the info message string to use if
+   *                       the optional field is missing. If not provided,
+   *                       a default info message will be generated.
+   */
+  std::string getOptionalField(const Node& node, const std::string& field_name, const std::string& default_value, const std::string& section = "", const std::string& info_message = "");
+
+  static std::shared_ptr<utils::IdGenerator> id_generator_;
+  std::unordered_set<std::string> uuids_;
+  std::shared_ptr<logging::Logger> logger_;
```

Comment:
I would argue that `JsonConfiguration` having a `StructuredConfiguration` base class is an implementation detail and all log messages should "originate" from the most derived class, e.g. having an error from `TailFile`'s base class `Processor`, through `TailFile`'s logger is easy to track, while having multiple messages from `Processor::logger_` we cannot be sure which processor it comes from (e.g. if we have both a `TailFile` and a `MergeContent`)

Reply:
this is kind of relevant, users ask for more specificity: https://issues.apache.org/jira/browse/MINIFICPP-1948

Reply:
I don't think this will give much extra info to users, since each minifi will either have a yaml or a json configuration but not both.  On the other hand, it will make our future selves' lives slightly harder.  But if you think it's worth it, I don't mind.